tests/mr/device/arena_mr_tests.cpp

```diff
@@ -15,48 +15,516 @@
+#include <rmm/detail/cuda_util.hpp>
 #include <rmm/detail/error.hpp>
 #include <rmm/mr/device/arena_memory_resource.hpp>
-#include <rmm/mr/device/cuda_memory_resource.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
 #include <rmm/mr/device/per_device_resource.hpp>
 
+#include <gmock/gmock.h>
 #include <gtest/gtest.h>
 
+#include <memory>
 #include <sys/stat.h>
 
 namespace rmm::test {
-
 namespace {
-using cuda_mr  = rmm::mr::cuda_memory_resource;
-using arena_mr = rmm::mr::arena_memory_resource<rmm::mr::cuda_memory_resource>;
 
-TEST(ArenaTest, ThrowOnNullUpstream)
+class mock_memory_resource {
+ public:
+  MOCK_METHOD(void*, allocate, (std::size_t));
+  MOCK_METHOD(void, deallocate, (void*, std::size_t));
+};
+
+using rmm::mr::detail::arena::block;
+using rmm::mr::detail::arena::byte_span;
+using rmm::mr::detail::arena::superblock;
+using global_arena = rmm::mr::detail::arena::global_arena<mock_memory_resource>;
+using arena        = rmm::mr::detail::arena::arena<mock_memory_resource>;
+using arena_mr     = rmm::mr::arena_memory_resource<rmm::mr::device_memory_resource>;
+using ::testing::Return;
+
+// NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast,performance-no-int-to-ptr)
+auto const fake_address  = reinterpret_cast<void*>(1_KiB);
+auto const fake_address2 = reinterpret_cast<void*>(2_KiB);
+auto const fake_address3 = reinterpret_cast<void*>(superblock::minimum_size);
+auto const fake_address4 = reinterpret_cast<void*>(superblock::minimum_size * 2);
+// NOLINTEND(cppcoreguidelines-pro-type-reinterpret-cast,performance-no-int-to-ptr)
+
+class ArenaTest : public ::testing::Test {
+ protected:
+  void SetUp() override
+  {
+    EXPECT_CALL(mock_, allocate(arena_size_)).WillOnce(Return(fake_address3));
+    EXPECT_CALL(mock_, deallocate(fake_address3, arena_size_));
+    global_arena_ = std::make_unique<global_arena>(&mock_, arena_size_);
+    arena_        = std::make_unique<arena>(*global_arena_);
+  }
+
+  // NOLINTBEGIN(cppcoreguidelines-non-private-member-variables-in-classes)
+  std::size_t arena_size_{superblock::minimum_size * 4};
+  mock_memory_resource mock_{};
+  std::unique_ptr<global_arena> global_arena_{};
+  std::unique_ptr<arena> arena_{};
+  // NOLINTEND(cppcoreguidelines-non-private-member-variables-in-classes)
+};
+
+/**
+ * Test align_to_size_class.
+ */
+TEST_F(ArenaTest, AlignToSizeClass)  // NOLINT
+{
+  using rmm::mr::detail::arena::align_to_size_class;
+  EXPECT_EQ(align_to_size_class(8), 256);
+  EXPECT_EQ(align_to_size_class(256), 256);
+  EXPECT_EQ(align_to_size_class(264), 512);
+  EXPECT_EQ(align_to_size_class(512), 512);
+  EXPECT_EQ(align_to_size_class(17_KiB), 20_KiB);
+  EXPECT_EQ(align_to_size_class(13_MiB), 14_MiB);
+  EXPECT_EQ(align_to_size_class(2500_MiB), 2560_MiB);
+  EXPECT_EQ(align_to_size_class(128_GiB), 128_GiB);
+  EXPECT_EQ(align_to_size_class(1_PiB), std::numeric_limits<std::size_t>::max());
+}
+
+/**
+ * Test byte_span.
+ */
+
+TEST_F(ArenaTest, ByteSpan)  // NOLINT
```

Comment:
BTW, I'm hoping we will get GTEST clang-tidy errors fixed, obviating the need for these NOLINTs. https://github.com/google/googletest/pull/3676